core/common/src/InstantParser.kt

```diff
@@ -0,0 +1,141 @@
+ */
+
+package kotlinx.datetime
+
+import kotlin.math.min
+import kotlin.math.pow
+
+internal fun parseInstantCommon(string: String): Instant = parseIsoString(string)
+
+/*
+ * The algorithm for parsing time and zone offset was adapted from
+ * https://github.com/square/moshi/blob/aea17e09bc6a3f9015d3de0e951923f1033d299e/adapters/src/main/java/com/squareup/moshi/adapters/Iso8601Utils.java
+ */
+private fun parseIsoString(isoString: String): Instant {
+    try {
+        val dateTimeSplit = isoString.split('T', ignoreCase = true)
+        if (dateTimeSplit.size != 2) {
+            throw DateTimeFormatException("ISO 8601 datetime must contain exactly one (T|t) delimiter.")
+        }
+        val localDate = LocalDate.parse(dateTimeSplit[0])
+
+        // Iso8601Utils.parse
+        val timePart = dateTimeSplit[1]
+        var offset = 0
+        val hour = parseInt(timePart, offset, offset + 2).also { offset += 2 }
+        if (checkOffset(timePart, offset, ':')) {
+            offset += 1
+        }
+        val minutes = parseInt(timePart, offset, offset + 2).also { offset += 2 }
+        if (checkOffset(timePart, offset, ':')) {
+            offset += 1
+        }
+
+        var seconds = 0
+        var nanosecond = 0
+        // seconds and fraction can be optional
+        if (timePart.length > offset) {
+            val c = timePart[offset]
+            if (c != 'Z' && c != 'z' && c != '+' && c != '-') {
+                seconds = parseInt(timePart, offset, offset + 2).also { offset += 2 }
+                if (seconds > 59 && seconds < 63) { // https://github.com/Kotlin/kotlinx-datetime/issues/5
+                    seconds = 59 // truncate up to 3 leap seconds
+                }
+                if (checkOffset(timePart, offset, '.')) {
+                    offset += 1
+                    val endOffset =
+                        indexOfNonDigit(timePart, offset + 1) // assume at least one digit
+                    val parseEndOffset =
+                        min(endOffset, offset + 9) // parse up to 9 digits
+                    val fraction = parseInt(timePart, offset, parseEndOffset)
+                    nanosecond = (10.0.pow(9 - (parseEndOffset - offset)) * fraction).toInt()
+                    offset = endOffset
+                }
+            }
+        }
+
+        // extract timezone
+        if (timePart.length <= offset) {
+            throw DateTimeFormatException("No time zone indicator in '$timePart'")
+        }
+        val timezone: TimeZone
+        val timezoneIndicator = timePart[offset]
+        if (timezoneIndicator == 'Z' || timezoneIndicator == 'z') {
+            timezone = TimeZone.UTC
+        } else if (timezoneIndicator == '+' || timezoneIndicator == '-') {
+            val timezoneOffset = timePart.substring(offset)
+            // 18-Jun-2015, tatu: Minor simplification, skip offset of "+0000"/"+00:00"
```

Comment:
These comments have little meaning without the commit history, so they probably shouldn't be included here.